DietPi-Drive_Manager | Added ability to adjust reserved blocks percentage on ext4 file system

MichaIng opened the pull request:
https://github.com/Fourdee/DietPi/issues/1662
+ DietPi-Drive_Manager | Minor code improvements

Commit:
v6.8

+ DietPi-Drive_Manager | Added ability to adjust reserved blocks percentage on ext4 file system
+ DietPi-Drive_Manager | Minor code improvements

Commit:
v6.8

+ DietPi-Drive_Manager | Added ability to adjust reserved blocks percentage on ext4 file system

Fourdee committed:
G_AG_CHECK_INSTALL_PREREQ rsync

Fourdee commented on dietpi/dietpi-drive_manager:
```diff
@@ -998,17 +1000,35 @@
+				G_WHIP_INPUTBOX 'Ext4 formatted drives allow the reservation of drive space for the root user to assure system functionality, if filled by other users or processes, and to avoid fragmentation of large files.\n\nHowever, on modern drives, the default of 5% reserved blocks is often by orders of magnitude larger than necessary. You may want to reduce the percentage to an absulte reserved space o ...
+				if (( ! $? )); then
+
+					if grep -qE '^[0-9]+\.?[0-9]*$' <<< "$G_WHIP_RETURNED_VALUE" &&
```
We should be aiming to use global here:
```
G_CHECK_VALIDINT
```

We should tweak ```G_CHECK_VALIDINT``` if needed?

MichaIng replied:
Maybe G_CKECK_VALIDNUM and allow a variable/option to allow floats as well, which is needed here?

Would be also nice if we find a more elegant way to compare float values, than sed it into integers :smile:.

Fourdee replied:
@MichaIng 

> float

Good point 👍 lol

```G_CHECK_VALIDFLOAT```?

MichaIng replied:
@Fourdee 
Jep possible as well, maybe separate checks are more secure/cleaner. Just to have reference, though the solutions are similar to mine: https://stackoverflow.com/a/13790836
- But will we need this more often, that a global function already make sense? 😉

MichaIng replied:
@Fourdee 
`grep -qE '^[0-9]*\.?[0-9]*$'` works fine as float test, as ".5" == "0.5" and "5." == "5.0" is also accepted.
But I am not sure if this is true for all commands that generally accept floats. If we want to create `G_CHECK_VALIDFLOAT`, then perhaps it's better to just allow values with filled zero from our end. 

Also:
```
root@VM-Stretch:~# bc -l <<< '05 == 5'
1
root@VM-Stretch:~# tune2fs -m 05 /dev/sda1
tune2fs 1.43.4 (31-Jan-2017)
Setting reserved blocks percentage to 5% (104844 blocks)
```

Leading zeros are allowed by check as above and just ignored by `bc` and `tune2fs`, but this might be not true as well for other commands. Quite difficult to test, if we don't want to allow leading zeros, and want to have it, if followed by decimal dot...

Fourdee reviewed (commented):
Minor ```G_CHECK_VALIDINT``` use?

Fourdee committed:
Merge branch 'testing' into reserved_blocks_adjustment

MichaIng commented:
- [x] Use `bc` for float value comparison: https://github.com/Fourdee/DietPi/issues/1510#issuecomment-381994727
To check range: `bc -l <<< "( 0 <= $value ) * ( $value <= 50 )"` or `bc -l <<< "( 0 <= $value ) && ( $value <= 50 )"`
Will do commit this evening.

Commit:
v6.8

+ DietPi-Drive_Manager | Use "bc" to compare float values
+ DietPi-Drive_Manager | Allow ".5" as float, which will be read as "0.5", same as "5." will be read as "5.0"
+ DietPi-Drive_Manager | Minor code improvements

MichaIng commented on dietpi/dietpi-drive_manager:
```diff
@@ -124,31 +120,25 @@
-			local fp_temp='/tmp/.dietpi-drive_manager_devlist'
-			blkid -o device > $fp_temp
-			readarray -t aDRIVE_MOUNT_SOURCE < $fp_temp
+			readarray -t aDRIVE_MOUNT_SOURCE <<< "$(blkid -o device)"
```
Use of here strings to feed array, avoid temp file creation, if just read as single time.

Fourdee replied:
Nice 👍

MichaIng commented on dietpi/dietpi-drive_manager:
```diff
@@ -124,31 +120,25 @@
+			readarray -t aDRIVE_FSTYPE <<< "$(blkid -s TYPE -o value)"
 
 			# - Check fstab for existing entries and mount locations
-			local mount_index=1
```
I don't find "$mount_index" used any more?

Fourdee replied:
Yep good spot 👍

MichaIng commented on dietpi/dietpi-drive_manager:
```diff
@@ -998,17 +994,35 @@
+
+					else
+
+						G_WHIP_MSG 'Error:\n\nPlease enter an integer or float between "0" and "50", e.g. "0.5" or "10".'
```
@Fourdee 
- [x] Do you know a good way to let user directly retry the input, if entry was found wrong? Hmm just have the idea to use a while loop und do break if float check was passed?

Fourdee replied:
@MichaIng 

> just have the idea to use while [ "$G_WHIP_RETURNED_VALUE" = 'Reserved blocks' ] loop und do break if float check was passed?

Yep spot on 👍

MichaIng replied:
@Fourdee 
Done, jep that's the way I like it 😃.

Commit:
v6.8

+ Whoopsie: (( $(bc ... ) )) needed

Commit:
v6.8

+ DietPi-Drive_Manager | Reserved blocks: Allow directly inputbox retry, if no valid number was found

Fourdee commented:
@MichaIng 

Looking good, nice work on the while loop 👍 

Lets merge for testing.

MichaIng commented:
@Fourdee 
Only possible error I found was entering just `.` into the field...
`[0-9]+\.` would take a bid flexibility (allowing functional `.5` for `0.5`), but assure that no wrong value will pass the test.
Dunno, who will enter just a dot 🤣.

Fourdee commented:
@MichaIng 

> Dunno, who will enter just a dot

Dot matrix printer enthusiast? 😄 

I'll send a commit that prevents ```.``` value. There is probably a better way of checking for this, but, will do for now.

MichaIng commented:
@Fourdee 
Jep, also thought about this solution, but was too shy to add the third condition check there 🤣. Jep at least this really guarantees full flexibility and no possible wrong value (as far as I can think of) 👍. We can think later about a beautiful check, when adding G_CHECK_VALIDFLOAT e.g.
